models/mobilenet/mobilenet_v1_impl.hpp

```diff
@@ -0,0 +1,146 @@
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0),
+    alpha(0),
+    depthMultiplier(0)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1(
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const size_t alpha,
+    const size_t depthMultiplier,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    MobileNetV1<OutputLayerType, InitializationRuleType>(
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        alpha,
+        depthMultiplier,
+        includeTop,
+        preTrained,
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+MobileNetV1<OutputLayerType, InitializationRuleType>::MobileNetV1(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const size_t alpha,
+    const size_t depthMultiplier,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    alpha(alpha),
+    depthMultiplier(depthMultiplier),
+    numClasses(numClasses)
+{
+  if (inputWidth < 32 || inputHeight < 32)
+  {
+    mlpack::Log::Fatal << "input width and input height cannot be smaller than" 
+        " 32.\nGiven input width and height: (" << inputWidth << ", "
+        << inputHeight << ")" << std::endl;
+  }
+  outSize = size_t(32 * alpha);
+  mobileNet.Add(new ann::Convolution<>(3, outSize, 3, 3, 2, 2, 0, 0,
+      inputWidth, inputHeight, "same"));
+  mlpack::Log::Info << "Convolution: " << "(" << "3, " << inputWidth << ", "
+      << inputHeight << ")" << " ---> (" << outSize << ", ";
+  inputWidth = ConvOutSize(inputWidth, 3, 2, 1);
+  inputHeight = ConvOutSize(inputHeight, 3, 2, 1);
+  mlpack::Log::Info << inputWidth << ", " << inputHeight << ")" << std::endl;
+  mobileNet.Add(new ann::BatchNorm<>(outSize, 1e-3, true, 0.99));
+  mlpack::Log::Info << "BatchNorm: " << "(" << outSize << ")"
+        << " ---> (" << outSize << ")" << std::endl;
+  ReLU6Layer();
+  outSize = DepthWiseConvBlock(outSize, 64, alpha, depthMultiplier);
+  
+  for (const auto& blockConfig: mobileNetConfig)
+  { 
+    outSize = DepthWiseConvBlock(outSize, blockConfig.first, alpha,
+        depthMultiplier, 2);
+
+    for(size_t numBlock = 1; numBlock < blockConfig.second; ++numBlock)
+    {
+      outSize = DepthWiseConvBlock(outSize, blockConfig.first, alpha,
+          depthMultiplier);
+    }
+
+  }
+
+  if (includeTop)
+  {
+    // We need something like Global average pooling, refernce:
+    // 1. https://keras.io/api/layers/pooling_layers/global_average_pooling2d/
```

Comment:
@zoq, @kartikdutt18 we probably also need something like global avg pooling, in this the conv layer generates equal generates a output of `[1, 1024, 7, 7]` in the order where `[batch, channel, width, height]` so a global avg pool layer takes the avg of `[1, 1->1024, 7, 7]` and produces a output of `[1, 1024]` we don't have anything like this right now, but maybe i am missing something so do let me know.

Reply:
Instead of adding fully connected layers on top of the feature maps, we take the average of each feature map, and the resulting vector is fed directly into the softmax layer. One advantage of global average pooling over the fully connected layers is that it is more native to the convolution structure by enforcing correspondences between feature maps and categories. Thus the feature maps can be easily interpreted as categories confidence maps. Another advantage is that there is no parameter to optimize in the global average pooling thus overfitting is avoided at this layer. Futhermore, global average pooling sums out the spatial information, thus it is more robust to spatial translations of the input.

A more formal defination if the one that i wrote confuses us. 
and aa proper reference: https://www.machinecurve.com/index.php/2020/01/30/what-are-max-pooling-average-pooling-global-max-pooling-and-global-average-pooling/#global-average-pooling

Reply:
Agreed, having global avg pooling is probably the cleanest way to implement this.

Reply:
I think we can use ApativePooling layer with output size (1,1). It should work same as global average / max pooling.

Reply:
But that will pool over the output to reduce it to what we want, rather than we want to just take the mean of that [7*7] block for every 1024 output we have.

Reply:
So the number of dims might be same but the way how it reduces it would change, right ?

Reply:
Could you take a look at this https://colab.research.google.com/drive/1g8ybtpHFCN_VIQWgi4zH4frciao_MUwD?usp=sharing
Because we do have average pooling layer in mlpack.

Reply:
makes sense, thanks for writing this. So i believe our adaptive avg pool with input as (1, 1) should work just fine.

Reply:
Yes.